### ISSUE TITLE
False positive `reportAssignmentType` with `list.__add__` type widening from a newly-created object

### ISSUE THREAD

**Issue by Phrogz** [DetachHead/basedpyright#1486]
### Description

```python
foo: list[str] = ["a", "b"]
bar: list[str | float] = foo + ["c"]     # incorrectly shows a type-widening error here
baz: list[str | float] = ["a", "b", "c"] # type-widening is correctly accepted here
bat: list[str | float] = foo             # correctly shows a type-widening error here
```

The `bar` line should be treated as the `baz`, as `list + list` creates a new list with no other references to it; mutating `bar` does not/cannot affect `foo`.

[Link to BasedPyright Playground Example](https://basedpyright.com/?typeCheckingMode=all&code=GYexC4AIBsEsGcAuBtJAnAupAvJZAiAQ3wBpJ8AjfDAKAsLSjiVUTUgB9JhoRDEsuUCEgBqPPgDG1IA)

Occurs in Pyright as well; filed upstream as [Pyright bug #10943](https://github.com/microsoft/pyright/issues/10943).

**Comment by KotlinIsland** [comment]
hmm, is this a typeshed issue?

**Comment by beauxq** [comment]
I think it would require extra steps of inference (likely a performance limitation)
or some kind of ownership model in the type system to know that there are no other references to the value returned by `list.__add__`

**Comment by KotlinIsland** [comment]
can you provide an example of this being unsound? as far as I can think:

```py
class list[T]:
    def __add__[O](self, other: list[O]) -> list[T | O]:
        result = []
        result.extend(self)
        result.extend(other)
        return result
```

this seems valid to me

**Comment by beauxq** [comment]
I don't think that's unsound.
But how are we supposed to get the inference of the type variables that we need from the code in the OP?

**Comment by KotlinIsland** [comment]
I haven't investigated the signature yet, but there is no issue inferring `other` as `list[float]`

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2Pvqja3qQCeAbQAqAXQBcjKDKjTZAEzTBYMLKSxI4ACnpo2wADRQkEqCICUUALQA%2BKADk6aKbLdRcnJvJlKVcYgUFOCEAeTFdfSMoSgALNBAzXn5hcIs7Hj5BUSgAHyhwiQA6EuZiJKzhfhAxKABeTJTtACJmi0YAIwqUoWq8qGA2MGpahuIoAGpGwW0ARiKABgsgA)

**Comment by beauxq** [comment]
> I haven't investigated the signature yet, but there is no issue inferring `other` as `list[float]`

The OP has `["c"]` in the position that would need to be `list[float]`

**Comment by KotlinIsland** [comment]
ah, my mistake, it's still normal bidirectional type inference, should be fine

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2Pvqja3qQCeAbQAqAXQBcjKDKjTZAEzTBYMLKSxI4ACnpo2wADRQkEqCICUUALQA%2BKADk6aKbLdRcnJvJlKVcYgUFOCEAeTFdfSMoSgALNBAzXn5hcIs7Hj5BUSgAHyhwiQA6EuZiJKzhfhAxKABeTJTtACJmi0YAIwqUoWq8qGA2MGpahuIoAGpGwRaWNqA)

**Comment by KotlinIsland** [comment]
```py
    # Overloading looks unnecessary, but is needed to work around complex mypy problems
    @overload
    def __add__(self, value: list[_T], /) -> list[_T]: ...
    @overload
    def __add__(self, value: list[_S], /) -> list[_S | _T]: ...
```

so this is a nonsense mypy issue, i think we update the vendored typeshed and see what comes out of the primer

**Comment by beauxq** [comment]
> ah, my mistake, it's still normal bidirectional type inference, should be fine
> 
> Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2Pvqja3qQCeAbQAqAXQBcjKDKjTZAEzTBYMLKSxI4ACnpo2wADRQkEqCICUUALQA%2BKADk6aKbLdRcnJvJlKVcYgUFOCEAeTFdfSMoSgALNBAzXn5hcIs7Hj5BUSgAHyhwiQA6EuZiJKzhfhAxKABeTJTtACJmi0YAIwqUoWq8qGA2MGpahuIoAGpGwRaWNqA)

That sample is covariant.

This one is invariant: [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2Pvqja3qQCeAbQAqAXQBcjKDKjTZAEzTBYMLKSxI4ACnpo2wADRQkEqCICUUALQA%2BKADk6aKbLdRcnJvJlKVxXFw0UgVdfSMTM0tXdxlcEHUkbSQLHzk3PygAczQtTTQIMIMrO3MY2IA6KuYM5VViBQU4IQB5MSKIygALNBAzXn5hNotSgcFRKAAfKDaJKormYn6%2Bcf4QMSgAXh4VgW0AIn3UgCNlwaE1qahgNjBqDe3iKABqHcGDliPGIA)

**Comment by DetachHead** [comment]
i think this needs more than just a typeshed update. there seems to be some special casing when using the `+` operator that results in different behavior to when the dunder is called like a regular method:

```diff
-     # Overloading looks unnecessary, but is needed to work around complex mypy problems
-     @overload
-     def __add__(self, value: list[_T], /) -> list[_T]: ...
-     @overload
-     def __add__(self, value: list[_S], /) -> list[_S | _T]: ...
+     def __add__(self, value: list[_S | _T], /) -> list[_S | _T]:
        """Return self+value."""
        ...
```

```py
foo: list[str] = ["a", "b"]
bar: list[str | float] = foo.__add__(["c"])  # now works
baz: list[str | float] = foo + ["c"] # still errors
```

**Comment by KotlinIsland** [comment]
i fixed it: Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2Pvqja3qQCeAbQAqAXQBcjKDKjTZAEzTBYMLKSxI4ACnpo2wADRQkEqCICUUALQA%2BKADk6aKbLdRcnJvJlKVxXFw0UgVdfSMTM0tXdxlcEHUkbSQLHzk3PygAczQtTTQIMIMrO3MY2IA6KuYM5VViBQU4IQB5YwBRMSKIygALNBAzXn5hFrELUuHBIXaoAB9zeagxiSqK5mIhvmn%2BEDEoAF4ebYFtACIz1IAjLZGhMCuAKzQWJH2j4igAamOR85ZLowgA)


we need to capture the expected type as well

**Comment by KotlinIsland** [comment]
extreme sus detected: Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2PvqgCEwAPAIwBtACoBdAFyMosqDLkATNMFgwspLEjgAKemjbAANFCSSoogJRQAtAD4oAOTpppc91H2GAdGYtQAXlNmd2VVOGJFRThhAHkTAFFxPQNjKEoACzQQcz4hOPFLe14BEQSoAB9-Ktipb3rmACNBGHokHJL8tpBxQM7BHQAiRsHLRmaYMEaAKzQWPzyRKdn53qCJ7qgAan6hllHmdk5uPIAmMSkFWTC1DS1dLzS-KzcPWUffc1E%2BpBClFTUkWiMDiiWSjxMmWyuQE51q1mKizilSg5SqEnM9W8TVOrXafTOQxGY0auOWcwWsOE5NWfVJeJA206pz2ByAA)


looks like the order of the union is affecting the behaviour

- #1492

**Comment by DetachHead** [comment]
> i think this needs more than just a typeshed update. there seems to be some special casing when using the `+` operator that results in different behavior to when the dunder is called like a regular method:
> 
> ```diff
> -     # Overloading looks unnecessary, but is needed to work around complex mypy problems
> -     @overload
> -     def __add__(self, value: list[_T], /) -> list[_T]: ...
> -     @overload
> -     def __add__(self, value: list[_S], /) -> list[_S | _T]: ...
> +     def __add__(self, value: list[_S | _T], /) -> list[_S | _T]:
>         """Return self+value."""
>         ...
> ```
> 
> ```py
> foo: list[str] = ["a", "b"]
> bar: list[str | float] = foo.__add__(["c"])  # now works
> baz: list[str | float] = foo + ["c"] # still errors
> ```
#1493
